### PR TITLE
feat: improved route broadcast

### PR DESCRIPTION
**Pull request by momerath42**
+ route versioning and per-connector filtering of redundancy
+ link-down detection and (re)broadcast

### CODE REVIEW

File: src/lib/message-router.js [hunk 1]
```diff
@@ -105,7 +106,25 @@ MessageRouter.prototype._handleRequest = function * (request, sender) {
...
+    for (const ledger of potentiallyUnreachableLedgers) {
+      lostLedgerLinks.concat(this.routingTables.invalidateConnectorsRoutesTo(sender, ledger))
+    }
+  }
+
+  if (!routes && (lostLedgerLinks.length === 0)) { // just a heartbeat
+    log.info('got heartbeat from:', sender)
+    return
+  }
   validate('Routes', routes)
```
Comment (sentientwaffle) [inline src/lib/message-router.js]: Schema validation should take place at the very beginning of this function (a new schema, since I don't see one that matches this format).

File: src/lib/message-router.js [hunk 2]
```diff
@@ -105,7 +106,25 @@ MessageRouter.prototype._handleRequest = function * (request, sender) {
...
+  let routes = payload.new_routes
+
+  let holdDownTime = payload.hold_down_time
+  this.routingTables.bumpConnector(sender, holdDownTime)
+  let potentiallyUnreachableLedgers = payload.unreachable_through_me
+  let lostLedgerLinks = []
+  if (potentiallyUnreachableLedgers.length > 0) {
+    log.info('informed of broken routes to:', potentiallyUnreachableLedgers, ' through:', sender)
+    for (const ledger of potentiallyUnreachableLedgers) {
+      lostLedgerLinks.concat(this.routingTables.invalidateConnectorsRoutesTo(sender, ledger))
```
Comment (sentientwaffle) [inline src/lib/message-router.js]: This is a noop; concat doesn't modify either array (and there should be tests to catch that as well).

File: src/lib/route-broadcaster.js [hunk 3]
```diff
@@ -27,6 +27,7 @@ class RouteBroadcaster {
     this.routeCleanupInterval = config.routeCleanupInterval
     this.routeBroadcastInterval = config.routeBroadcastInterval
     this.routingTables = routingTables
+    if (this.routingTables.publicTables.currentEpoch !== 0) throw new Error('expecting a fresh routingTables with epoch support')
```
Comment (sentientwaffle) [inline src/lib/route-broadcaster.js]: Why is this needed?

File: src/lib/route-broadcaster.js [hunk 4]
```diff
@@ -60,21 +72,33 @@ class RouteBroadcaster {
...
         log.warn('removing expired routes failed')
         log.debug(err)
       }
-      this.removeExpiredRoutesSoon()
     }, this.routeCleanupInterval)
   }
 
+  markLedgersUnreachable (lostLedgerLinks) {
+    if (lostLedgerLinks.length > 0) log.info('detected lostLedgerLinks:', lostLedgerLinks)
+    lostLedgerLinks.map((unreachableLedger) => { this.detectedDown.add(unreachableLedger) })
```
Comment (sentientwaffle) [inline src/lib/route-broadcaster.js]: Since nothing is being returned, use `forEach`.

File: src/lib/route-broadcaster.js [hunk 5]
```diff
@@ -60,21 +72,33 @@ class RouteBroadcaster {
...
+  _endEpoch () {
+    this.routingTables.publicTables.incrementEpoch()
+  }
+
   broadcastSoon () {
     defer.setTimeout(function * () {
       try {
         this.routingTables.removeExpiredRoutes()
-        yield this.reloadLocalRoutes()
+        yield this.reloadLocalRoutes() // todo? remove
```
Comment (sentientwaffle) [inline src/lib/route-broadcaster.js]: Why would this be removed?

File: test/quoteSpec.js [hunk 6]
```diff
@@ -504,7 +504,8 @@ describe('Quotes', function () {
       yield assert.isRejected(quotePromise, InvalidBodyError, 'Missing required parameter: destination_address')
     })
 
-    it('quotes a multi-hop route', function * () {
+    // the functionality this tests has been disabled, maybe only temporarily
+    xit('quotes a multi-hop route', function * () {
```
Comment (sentientwaffle) [inline test/quoteSpec.js]: `it.skip` is more clear

File: src/lib/message-router.js (hunk 2 above)
Comment (momerath42) [inline src/lib/message-router.js]: doh! good call - should have caught this after the bug in ilp-routing

File: src/lib/message-router.js (hunk 1 above)
Comment (momerath42) [inline src/lib/message-router.js]: I added schema validation, but the validation is failing and I don't understand why.  The call stack is vague in the extreme, and lots of manual comparison isn't clueing me in:

  1) RouteBroadcaster broadcast invalidates routes:
     InvalidBodyError: RoutingUpdate schema validation error: should be string
      at validate (src/lib/validate.js:18:11)
      at MessageRouter.receiveRoutes (src/lib/message-router.js:111:3)
      at next (native)
      at onFulfilled (node_modules/co/index.js:65:19)
      at node_modules/co/index.js:54:5
      at Context.co (node_modules/co/index.js:50:10)
      at Context.toPromise (node_modules/co/index.js:118:63)
      at next (node_modules/co/index.js:99:29)
      at onFulfilled (node_modules/co/index.js:69:7)
      at node_modules/co/index.js:54:5
      at Context.co (node_modules/co/index.js:50:10)
      at Context.createPromise (node_modules/co/index.js:30:15)
      at Test.Runnable.run (node_modules/co-mocha/lib/co-mocha.js:43:16)

the value that's failing: 
{ new_routes:
   [ { source_ledger: 'eur-ledger.',
       destination_ledger: 'random-ledger.',
       min_message_window: 1,
       source_account: 'eur-ledger.mary',
       points: [Object] } ],
  hold_down_time: 1234,
  unreachable_through_me: [] }

File: src/lib/route-broadcaster.js (hunk 3 above)
Comment (momerath42) [inline src/lib/route-broadcaster.js]: It was originally a way of making sure I was getting all the entry points into the related code.  Now, it serves to error out early if you try to use an incompatible ilp-routing, which has been helpful to me.  Its use may be served now though.

File: src/lib/route-broadcaster.js (hunk 5 above)
Comment (momerath42) [inline src/lib/route-broadcaster.js]: I guess it wouldn't, judging by your response (that's why there was a question mark).  I thought at the time that they might be reloaded because they expired, along with everything else.

File: src/lib/message-router.js (hunk 1 above)
Comment (sentientwaffle) [inline src/lib/message-router.js]: Your schema specifies `hold_down_time` as a [`NonNegativeDuration.json`](https://github.com/interledgerjs/five-bells-shared/blob/master/schemas/NonNegativeDuration.json), which is a subset of `string`.

File: package.json [hunk 7]
```diff
@@ -39,7 +39,7 @@
     "five-bells-shared": "^21.2.0",
     "ilp-core": "~11.1.0",
```
Comment (sentientwaffle) [inline package.json]: Note that this needs to be upgraded to https://github.com/interledgerjs/ilp-core/pull/79 before merging.

File: src/lib/message-router.js [hunk 8]
```diff
@@ -105,18 +106,39 @@ MessageRouter.prototype._handleRequest = function * (request, sender) {
...
+  let potentiallyUnreachableLedgers = payload.unreachable_through_me
+  let lostLedgerLinks = []
+  if (potentiallyUnreachableLedgers.length > 0) {
+    log.info('informed of broken routes to:', potentiallyUnreachableLedgers, ' through:', sender)
+    for (const ledger of potentiallyUnreachableLedgers) {
+      lostLedgerLinks.push(...this.routingTables.invalidateConnectorsRoutesTo(sender, ledger))
+    }
+  }
 
+  if (!routes && (lostLedgerLinks.length === 0)) { // just a heartbeat
```
Comment (sentientwaffle) [inline src/lib/message-router.js]: When would `!routes` ever trigger? `payload` is a `RoutingUpdate`, which means `payload.new_routes` is always an `Array`.

File: src/lib/route-broadcaster.js [hunk 9]
```diff
@@ -86,27 +110,37 @@ class RouteBroadcaster {
...
   }
 
-  _broadcastToLedger (adjacentLedger, routes) {
+  _broadcastToLedger (adjacentLedger, routes, unreachableLedgers) {
     const connectors = Object.keys(this.peersByLedger[adjacentLedger])
     return Promise.all(connectors.map((account) => {
       log.info('broadcasting ' + routes.length + ' routes to ' + account)
-
+      let routesNewToConnector = routes.filter((route) => (route.added_during_epoch > (this.peerEpochs[account] || -1)))
+      routesNewToConnector.map((r) => delete r.added_during_epoch)
```
Comment (sentientwaffle) [inline src/lib/route-broadcaster.js]: `forEach`, not `map`

File: schemas/RoutingUpdate.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    "unreachable_through_me": {
+      "description": "A list of ledgers that have become unreachable through the sending connector",
+      "type": "array",
+      "minItems": 0,
+      "items": {
+        "$ref": "IlpAddress.json"
+      }
+    },
+    "new_routes": {
+      "description": "A list of Routes that have been added to the sending connector's table since the last update it has sent you",
```
Comment (justmoon) [inline schemas/RoutingUpdate.json]: Might be good to mention here that the new routes could replace existing routes?
